// chime-sma-translator/dist/sma_mapping_actions/invoke-lambda.js
"use strict";
/*
Copyright (c) 2023 tekVizion PVS, Inc.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.InvokeLambda = void 0;
const call_details_1 = require("../utility/call-details");
const constant_values_1 = require("../const/constant-values");
const default_termination_action_1 = require("../utility/default-termination-action");
const error_types_1 = require("../const/error-types");
const contact_flow_processor_1 = require("../contact-flow-processor");
const next_action_error_handler_1 = require("../utility/next-action-error-handler");
const aws_sdk_1 = require("aws-sdk");
const metric_updation_1 = require("../utility/metric-updation");
/**
  * Invokes the External Lambda Function and stores the result in the ContextStore
  * @param smaEvent
  * @param action
  * @param actions
  * @param amazonConnectInstanceID
  * @param bucketName
  * @param contextStore
  * @returns The Next SMA action to perform
  */
class InvokeLambda {
    async processFlowActionInvokeLambdaFunction(smaEvent, action, actions, amazonConnectInstanceID, bucketName, contextStore) {
        let callId;
        let regionVal = constant_values_1.Attributes.region;
        console.log("the Region Value is " + regionVal);
        //creating the lambda API to invoke from the specific region
        const lambda = new aws_sdk_1.Lambda({ region: regionVal });
        // creating cloud watch metric parameter and updating the metric details in cloud watch
        let metric = new metric_updation_1.CloudWatchMetric();
        let params1 = metric.createParams(contextStore, smaEvent);
        console.log(constant_values_1.Attributes.DEFAULT_LOGGER + callId + " Invoke Lombda Action:");
        try {
            // getting the CallID of the Active call from the SMA Event
            let callDetails = new call_details_1.CallDetailsUtil();
            const legA = callDetails.getLegACallDetails(smaEvent);
            callId = legA.CallId;
            if (!callId)
                callId = smaEvent.ActionData.Parameters.CallId;
            let LambdaARN = action.Parameters.LambdaFunctionARN;
            // getting the input parameter form the Contact flow for invoking the external Lambda function
            let inputForInvoking = await inputForInvokingLambda(action, contextStore);
            console.log(constant_values_1.Attributes.DEFAULT_LOGGER + callId + " Input for invoking lambda Function" + JSON.stringify(inputForInvoking));
            const params = {
                FunctionName: LambdaARN,
                InvocationType: 'RequestResponse',
                Payload: JSON.stringify(inputForInvoking) //Mandatory
            };
            // invoking the external lambda function with the parameters and getting the result
            let result = await lambda.invoke(params).promise();
            console.log(constant_values_1.Attributes.DEFAULT_LOGGER + callId + " Invoke Lombda Action Result is " + result);
            if (!result) {
                params1.MetricData[0].MetricName = "InvokeLambdaNoResponse";
                metric.updateMetric(params1);
                let nextAction = await new next_action_error_handler_1.NextActionValidationUtil().getNextActionForError(action, actions, error_types_1.ErrorTypes.NO_MATCHING_ERROR, smaEvent);
                return await (0, contact_flow_processor_1.processFlowAction)(smaEvent, nextAction, actions, amazonConnectInstanceID, bucketName, contextStore);
            }
            params1.MetricData[0].MetricName = "InvokeLambdaSuccess";
            metric.updateMetric(params1);
            // parsing and storing the response after invoking the Lambda function 
            let x = JSON.parse(result.Payload.toString());
            console.log(constant_values_1.Attributes.DEFAULT_LOGGER + callId + " The Result After Invoking Lambda is" + JSON.stringify(x));
            // iterate the lambda response and store it in the contextStore
            const keys = Object.keys(x);
            keys.forEach((key, index) => {
                contextStore[constant_values_1.ContextStore.CONTEXT_ATTRIBUTES]["$.External." + key] = x[key];
            });
            // getting the next action object to execute
            let nextAction = callDetails.findActionObjectByID(actions, action.Transitions.NextAction);
            console.log(constant_values_1.Attributes.DEFAULT_LOGGER + callId + " Next Action identifier:" + action.Transitions.NextAction);
            return await (0, contact_flow_processor_1.processFlowAction)(smaEvent, nextAction, actions, amazonConnectInstanceID, bucketName, contextStore);
        }
        catch (error) {
            params1.MetricData[0].MetricName = "InvokeLambdaFailure";
            metric.updateMetric(params1);
            console.error(constant_values_1.Attributes.DEFAULT_LOGGER + callId + " There is an error in execution InvokeLambda" + error.message);
            return await new default_termination_action_1.TerminatingFlowUtil().terminatingFlowAction(smaEvent, "error");
        }
    }
}
exports.InvokeLambda = InvokeLambda;
/**
  * Get the inputs for invoking the Lambda from the Contact Flow action object
  * @param action
  * @param contextStore
  * @returns Process Flow Action
  */
async function inputForInvokingLambda(action, contextStore) {
    let InvocationAttributes = Object.entries(action.Parameters.LambdaInvocationAttributes);
    let contextAttributes = contextStore[constant_values_1.ContextStore.CONTEXT_ATTRIBUTES];
    for (let i = 0; i < InvocationAttributes.length; i++) {
        // checking if the attribute value contains any user defined, system or External attributes for replacing it to the corresponding value
        if (InvocationAttributes[i][1].includes("$.External.") || InvocationAttributes[i][1].includes("$.Attributes.")) {
            const keys = Object.keys(contextAttributes);
            keys.forEach((key, index) => {
                if (InvocationAttributes[i][1] == key)
                    InvocationAttributes[i][1] = InvocationAttributes[i][1].replace(key, contextAttributes[key]);
            });
        }
    }
    let lambdaFunctionParameters = Object.fromEntries(InvocationAttributes.map(([k, v]) => [k, v]));
    let inputForInvoking = {
        "Details": {
            "ContactData": {
                "Attributes": {},
                [constant_values_1.LambdaFunctionParameters.CHANNEL]: contextAttributes[constant_values_1.ContextAttributes.CHANNEL],
                [constant_values_1.LambdaFunctionParameters.CONTACTID]: contextAttributes[constant_values_1.ContextAttributes.CONTACTID],
                [constant_values_1.LambdaFunctionParameters.CUSTOMER_ENDPOINT]: {
                    [constant_values_1.LambdaFunctionParameters.ADDRESS]: contextAttributes[constant_values_1.ContextAttributes.CUSTOMER_ENDPOINT_ADDRESS],
                    [constant_values_1.LambdaFunctionParameters.TYPE]: contextAttributes[constant_values_1.ContextAttributes.CUSTOMER_ENDPOINT_TYPE]
                },
                [constant_values_1.LambdaFunctionParameters.INITIAL_CONTACTID]: contextAttributes[constant_values_1.ContextAttributes.INSTANCE_ARN],
                [constant_values_1.LambdaFunctionParameters.INITIATION_METHOD]: contextAttributes[constant_values_1.ContextAttributes.INITIATION_METHOD],
                [constant_values_1.LambdaFunctionParameters.INSTANCE_ARN]: contextAttributes[constant_values_1.ContextAttributes.INSTANCE_ARN],
                [constant_values_1.LambdaFunctionParameters.SYSTEM_ENDPOINT]: {
                    [constant_values_1.LambdaFunctionParameters.ADDRESS]: contextAttributes[constant_values_1.ContextAttributes.SYSTEM_ENDPOINT_ADDRESS],
                    [constant_values_1.LambdaFunctionParameters.TYPE]: contextAttributes[constant_values_1.ContextAttributes.SYSTEM_ENDPOINT_TYPE]
                }
            },
            "Parameters": lambdaFunctionParameters
        },
        "Name": "ContactFlowEvent"
    };
    return inputForInvoking;
}
