// chime-sma-translator/dist/sma_mapping_actions/lex-bot.js
/*
Copyright (c) 2023 tekVizion PVS, Inc. 

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LexBot = void 0;
const call_details_1 = require("../utility/call-details");
const chime_action_types_1 = require("../const/chime-action-types");
const constant_values_1 = require("../const/constant-values");
const default_termination_action_1 = require("../utility/default-termination-action");
const metric_updation_1 = require("../utility/metric-updation");
/**
  * Making a SMA action to perform delvier a Chat/Voice message and obtain customer input.
  * @param smaEvent
  * @param action
  * @param contextStore
  * @returns SMA action
  */
class LexBot {
    async processFlowActionConnectParticipantWithLexBot(smaEvent, action, contextStore) {
        let smaAction;
        let smaAction1;
        let callId;
        // creating cloud watch metric parameter and updating the metric details in cloud watch
        let metric = new metric_updation_1.CloudWatchMetric();
        let params = metric.createParams(contextStore, smaEvent);
        try {
            // getting the CallID of the Active call from the SMA Event
            let callDetails = new call_details_1.CallDetailsUtil();
            const legA = callDetails.getLegACallDetails(smaEvent);
            callId = legA.CallId;
            if (!callId)
                callId = smaEvent.ActionData.Parameters.CallId;
            console.log(constant_values_1.Attributes.DEFAULT_LOGGER + callId + " Start Bot Conversation");
            // checking user specified any lex session attributes in the Contact Flow
            if (action.Parameters.hasOwnProperty("LexSessionAttributes")) {
                smaAction = {
                    Type: chime_action_types_1.ChimeActions.START_BOT_CONVERSATION,
                    Parameters: {
                        BotAliasArn: action.Parameters.LexV2Bot.AliasArn,
                        LocaleId: 'en_US',
                        Configuration: {
                            SessionState: {
                                SessionAttributes: action.Parameters.LexSessionAttributes,
                                DialogAction: {
                                    Type: constant_values_1.Attributes.DIALOG_TYPE
                                }
                            },
                            WelcomeMessages: [
                                {
                                    ContentType: constant_values_1.Attributes.CONTENT_TYPE,
                                    Content: action.Parameters.Text
                                },
                            ]
                        }
                    }
                };
            }
            else {
                smaAction = {
                    Type: chime_action_types_1.ChimeActions.START_BOT_CONVERSATION,
                    Parameters: {
                        BotAliasArn: action.Parameters.LexV2Bot.AliasArn,
                        LocaleId: 'en_US',
                        Configuration: {
                            SessionState: {
                                DialogAction: {
                                    Type: constant_values_1.Attributes.DIALOG_TYPE
                                }
                            },
                            WelcomeMessages: [
                                {
                                    ContentType: constant_values_1.Attributes.CONTENT_TYPE,
                                    Content: action.Parameters.Text
                                },
                            ]
                        }
                    }
                };
            }
            let pauseAction = contextStore[constant_values_1.ContextStore.PAUSE_ACTION];
            params.MetricData[0].MetricName = "LexBotSuccess";
            metric.updateMetric(params);
            // checking if the pause action is there to perform before the actual action
            if (pauseAction) {
                smaAction1 = pauseAction;
                contextStore[constant_values_1.ContextStore.PAUSE_ACTION] = null;
                return {
                    "SchemaVersion": constant_values_1.Attributes.SCHEMA_VERSION,
                    "Actions": [
                        smaAction1, smaAction
                    ],
                    "TransactionAttributes": {
                        [constant_values_1.Attributes.CURRENT_FLOW_BLOCK]: action,
                        [constant_values_1.Attributes.CONNECT_CONTEXT_STORE]: contextStore
                    }
                };
            }
            return {
                "SchemaVersion": constant_values_1.Attributes.SCHEMA_VERSION,
                "Actions": [
                    smaAction
                ],
                "TransactionAttributes": {
                    [constant_values_1.Attributes.CURRENT_FLOW_BLOCK]: action,
                    [constant_values_1.Attributes.CONNECT_CONTEXT_STORE]: contextStore
                }
            };
        }
        catch (error) {
            params.MetricData[0].MetricName = "LexBotFailure";
            metric.updateMetric(params);
            console.error(constant_values_1.Attributes.DEFAULT_LOGGER + callId + " There is an error in execution of ConnectParticipantWithLexBot " + error.message);
            return await new default_termination_action_1.TerminatingFlowUtil().terminatingFlowAction(smaEvent, "error");
        }
    }
}
exports.LexBot = LexBot;
