// chime-sma-translator/dist/const/amazon-connect-action-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AmazonConnectActions = void 0;
// Amazon connect ations supported by the tekVizion's Library
var AmazonConnectActions;
(function (AmazonConnectActions) {
    AmazonConnectActions["GET_PARTICIPANT_INPUT"] = "GetParticipantInput";
    AmazonConnectActions["MESSAGE_PARTICIPANT"] = "MessageParticipant";
    AmazonConnectActions["DISCONNECT_PARTICIPANT"] = "DisconnectParticipant";
    AmazonConnectActions["WAIT"] = "Wait";
    AmazonConnectActions["UPDATE_CONTACT_RECORDING_BEHAVIOUR"] = "UpdateContactRecordingBehavior";
    AmazonConnectActions["LOOP"] = "Loop";
    AmazonConnectActions["TRANSFER_PARTICIPANT_TO_THIRD_PARTY"] = "TransferParticipantToThirdParty";
    AmazonConnectActions["CONNECT_PARTICIPANT_WITH_LEX_BOT"] = "ConnectParticipantWithLexBot";
    AmazonConnectActions["TRANSFER_TO_FLOW"] = "TransferToFlow";
    AmazonConnectActions["UPDATE_CONTACT_TEXT_TO_SPEECH"] = "UpdateContactTextToSpeechVoice";
    AmazonConnectActions["INVOKE_LAMBDA_FUNCTION"] = "InvokeLambdaFunction";
    AmazonConnectActions["UPDATE_CONTACT_ATTRIBUTES"] = "UpdateContactAttributes";
    AmazonConnectActions["COMPARE"] = "Compare";
    AmazonConnectActions["INVOKE_FLOW_MODULE"] = "InvokeFlowModule";
    AmazonConnectActions["END_FLOW_MODULE_EXECUTION"] = "EndFlowModuleExecution";
    AmazonConnectActions["DISTRIBUTE_BY_PERCENTAGE"] = "DistributeByPercentage";
    AmazonConnectActions["UPDATE_FLOW_LOGGING_BEHAVIOUR"] = "UpdateFlowLoggingBehavior";
})(AmazonConnectActions = exports.AmazonConnectActions || (exports.AmazonConnectActions = {}));
